internal/backends/postgres/postgres.go

```diff
@@ -0,0 +1,266 @@
+package postgres
+
+import (
+	"bytes"
+	"context"
+	"database/sql/driver"
+	_ "embed" // Needed for file embedding
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+)
+
+//go:embed schema.sql
+var schema string
+
+// Backend holds the state of the backend.
+//
+// Has to be initializes with New().
+type Backend struct {
+	pool *pgxpool.Pool
+}
+
+// New creates a new connection pool.
+func New(ctx context.Context, url string) (*Backend, error) {
+	conf, err := pgxpool.ParseConfig(url)
+	if err != nil {
+		return nil, fmt.Errorf("invalid connection url: %w", err)
+	}
+	conf.LazyConnect = true
+	pool, err := pgxpool.ConnectConfig(ctx, conf)
+	if err != nil {
+		return nil, fmt.Errorf("creating connection pool: %w", err)
+	}
+
+	b := Backend{
+		pool: pool,
+	}
+
+	return &b, nil
+}
+
+// Wait blocks until a connection to postgres can be established.
+func (b *Backend) Wait(ctx context.Context, log func(format string, a ...interface{})) {
+	for ctx.Err() == nil {
+		err := b.pool.Ping(ctx)
+		if err == nil {
+			return
+		}
+		if log != nil {
+			log("Waiting for postgres: %v", err)
+		}
+		time.Sleep(500 * time.Millisecond)
+	}
+}
+
+// Migrate creates the database schema.
+func (b *Backend) Migrate(ctx context.Context) error {
+	if _, err := b.pool.Exec(ctx, schema); err != nil {
+		return fmt.Errorf("creating schema: %w", err)
+	}
+	return nil
+}
+
+// Close closes all connections. It blocks, until all connection are closed.
+func (b *Backend) Close() {
+	b.pool.Close()
+}
+
+// Start starts a poll.
+func (b *Backend) Start(ctx context.Context, pollID int) error {
+	sql := `
+	INSERT INTO poll (poll_id, stopped) VALUES ($1, false) ON CONFLICT DO NOTHING;
+	`
+	if _, err := b.pool.Exec(ctx, sql, pollID); err != nil {
+		return fmt.Errorf("insert poll: %w", err)
+	}
+	return nil
+}
+
+// Vote adds a vote to a poll.
+func (b *Backend) Vote(ctx context.Context, pollID int, userID int, object []byte) error {
+	err := b.pool.BeginTxFunc(
+		ctx,
+		pgx.TxOptions{
+			IsoLevel: "REPEATABLE READ",
```

Comment:
It would be nice to comment in `New`, if autocommit is enabled or not. Maybe explicitly set the value you need and do not depend on the default of the postgres adapter.

Also a short sentence about transaction management by the adapter: When doing `QueryRow`/`Exec` below, is it wrapped in a transaction? From reading the code I think it is not. If it is not the case, I'm not sure, if the isolation level works as intended.

Reply:
I am not sure if I understand correctly.

I don't know what autocommit is. When I understand it correctly, then I think, that it is enabled. But I am not sure. I could not find this term in the library that I am using.

`b.poolBeginTxFunc` handels a transaction. See https://pkg.go.dev/github.com/jackc/pgx/v4#Conn.BeginTxFunc

Does this answers you questions?

Reply:
Ah I totally overlooked that `BeginTxFunc` starts a transaction.